core/src/main/scala/chisel3/Module.scala

```diff
@@ -283,6 +283,60 @@ abstract class Module extends RawModule {
   }
 }
 
+/** Provides an implicit Clock for use _within_ the [[RawModule]]
+  *
+  * Be careful to define the Clock value before trying to use it.
+  * Due to Scala initialization order, the actual val defining the Clock must occur before any
+  * uses of the implicit Clock.
+  *
+  * @example
+  * {{{
+  * class MyModule extends RawModule with ImplicitClock {
+  *   // Define a Clock value, it need not be called "implicitClock"
+  *   val clk = IO(Input(Clock()))
+  *   // Implement the virtual method to tell Chisel about this Clock value
+  *   override protected def implicitClock = clk
+  *   // Now we have a Clock to use in this RawModule
+  *   val reg = Reg(UInt(8.W))
+  * }
+  * }}}
+  */
+trait ImplicitClock { self: RawModule =>
+
+  /** Method that should point to the user-defined Clock */
+  protected def implicitClock: Clock
+
+  Builder.currentClock = Some(Delayed(implicitClock))
+}
+
+/** Provides an implicit Reset for use _within_ the [[RawModule]]
+  *
+  * Be careful to define the Reset value before trying to use it.
+  * Due to Scala initialization order, the actual val defining the Reset must occur before any
```

Comment:
nit: You say "the actual val" but then below you use a `def`. I can certainly imagine someone doing `override protected def implicitClock = InstantiateAClockGate(...)` and being very sad by all the clock gates getting put down. Should we use a `val` in the example body?

Reply:
The point I'm trying to illustrate is that the actual reset object needs to be a value, a `val`, but `implicitReset` can be a def. I could simplify the example a little bit by glossing over this subtly, but I was trying to show this point. What do you think I should do here? I could make this example very simple and put the more subtle example in a cookbook example.

Reply:
I could also just add more text explaining this fact.

Reply:
I've decided to keep the def but add a clarifying comment